resources/tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/internal/alias/BasicAliasTest.java

```diff
@@ -410,6 +410,8 @@ public void testBug256837() throws CoreException {
 		// change the location of p2 project to the temp folder
 		replaceProject(p2, tempStore.toURI());
 
+		waitForRefresh(); // takes some milliseconds until aliasManager is notified
```

Comment:
What is the contract of AliasManager? Is it supposed to be async?
I'm afraid this test failure shows we actually have something broken in the synchronization.

Reply:
I don't know. If the AliasManager has any contract it is for sure broken multiple times: It does not even try to be thread safe and is NOT synchronized anywhere. There is totally no guarantee from JDK that it works.

what happens is that AliasManager.resourceChanged(IResourceChangeEvent) is called asynchronously and marks AliasManager.changedProjects = true . With a bit luck of having x86 memory architecture the main thread will see the change ;-)

This "fix" is the same as already  done for 
testBug232426, testBug358830, testCreateFolderInBackground, BasicAliasTest.testDeepLink, ...
which all suffer from the same problem.

I'd be happy if someone makes AliasManager thread safe + synchron.

Reply:
> what happens is that AliasManager.resourceChanged(IResourceChangeEvent) is called asynchronously

That actually surprises me, I was kind of expecting POST_CHANGE resource listeners to happen synchronously. But maybe I build wrong assumption based on lucky side-effects all those years.
Looking at the code, I do not see where the resourceChanged() invocation becomes async. I see

Project.move() directly calls
Workspace.endOperation(), direcly calls
Workspace.broadcastPostChange(), directly calls
NotificationManager.notificationManager.broadcastChanges(tree, event, true), directly calls
NotificationManager.notify(), directly calls
IResourceChangeListener.resourceChanged(), and for AliasManager, directly changes the value of `changedProjects`, which is then used by AliasManager.computeAliases()

So can you please enlighten as to why this AliasManager would react asynchronously here? I'm missing something.

Reply:
I do not know WHY ... i am already happy that after much debugging i found that it does - if and only if the test fails  - which is was hard to reproduce (delay first ResourceException.provideStackTrace in main thread).

Something runs in parallel and its a race condition

Reply:
First thing I do is usually put `Thread.dumpStack()` into method where I' curious why and from where it is called

```
synchronized(Object.class) {
  Sysout(Thread.currentThread().getName());
  Thread.dumpStack();
}
```

Reply:
BTW you do not need to change the source for such but can put it as condition in a breakpoint.